source/bringauto/communication/fleet_protocol/FleetProtocol.cpp

```diff
@@ -125,26 +129,34 @@ bool FleetProtocol::receiveCommand() {
 
 void FleetProtocol::processBufferData(const buffer &bufferData) {
 	Command command;
-	MissionModule::AutonomyCommand protoCommand {};
+	json commandJson;
+	char *data = static_cast<char*>(bufferData.data);
 
-	if(!protoCommand.ParseFromArray(bufferData.data, bufferData.size_in_bytes)) {
+	try {
+		commandJson = json::parse(data, data + bufferData.size_in_bytes);
+		settings::Logger::logInfo("Parsing command: {}", commandJson.dump());
+	} catch (json::parse_error &) {
 		settings::Logger::logWarning("Cannot parse car command!");
 		return;
 	}
-	settings::Logger::logInfo("Parsing command: {}", protoCommand.ShortDebugString());
-
-	command.setAction(common_utils::EnumUtils::valueToEnum<EAutonomyAction>(protoCommand.action()));
 
 	std::vector<osm::Route::Station> missions;
-	for(const auto &messageStop: protoCommand.stops()) {
-		osm::Route::Station station;
-		station.name = messageStop.name();
-		station.latitude = messageStop.position().latitude();
-		station.longitude = messageStop.position().longitude();
-		missions.push_back(station);
+
+	try {
+		for(int i = 0; i < commandJson.at("stops").size(); i++) {
+			osm::Route::Station station;
+			station.name = commandJson.at("stops").at(i).at("name");
+			station.latitude = commandJson.at("stops").at(i).at("position").at("latitude");
+			station.longitude = commandJson.at("stops").at(i).at("position").at("longitude");
+			missions.push_back(station);
+		}
+		command.setMission(missions);
+		command.setRoute(commandJson.at("route"));
+		command.setAction(common_utils::EnumUtils::valueToEnum<EAutonomyAction>(std::string(commandJson.at("action"))));
+	} catch (json::exception &) {
```

Comment:
Command are being validated before they are sent to the virtual vehicle. Also the suggested changes are pointless because the "at" function also throws an exception if it cant find a key.